### ISSUE TITLE
KeyError: 'y'

### ISSUE THREAD

**Issue by bradleyswilson** [gboeing/osmnx#323]
Greetings. Appreciate the great package. I'm struggling to graph_from_file to run, getting a key error (see Traceback below). Running osmnx in a clean conda environment with strict channel priority as recommended by the install docs. 

The .osm file is the raw New Zealand Geofrabrik download, clipped with osmconvert to a bounding box. Any thoughts for troubleshooting?

```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-6-cc08ebb90c6e> in <module>
----> 1 G = ox.graph_from_file("kaikoura_clip.osm")

~/anaconda3/envs/ox/lib/python3.7/site-packages/osmnx/core.py in graph_from_file(filename, bidirectional, simplify, retain_all, name)
   1909     # create graph using this response JSON
   1910     G = create_graph(response_jsons, bidirectional=bidirectional,
-> 1911                      retain_all=retain_all, name=name)
   1912 
   1913     # simplify the graph topology as the last step.

~/anaconda3/envs/ox/lib/python3.7/site-packages/osmnx/core.py in create_graph(response_jsons, name, retain_all, bidirectional)
   1366     # use as weight
   1367     if len(G.edges) > 0:
-> 1368         G = add_edge_lengths(G)
   1369 
   1370     return G

~/anaconda3/envs/ox/lib/python3.7/site-packages/osmnx/core.py in add_edge_lengths(G)
   1199     # first load all the edges' origin and destination coordinates as a
   1200     # dataframe indexed by u, v, key
-> 1201     coords = np.array([[u, v, k, G.nodes[u]['y'], G.nodes[u]['x'], G.nodes[v]['y'], G.nodes[v]['x']] for u, v, k in G.edges(keys=True)])
   1202     df_coords = pd.DataFrame(coords, columns=['u', 'v', 'k', 'u_y', 'u_x', 'v_y', 'v_x'])
   1203     df_coords[['u', 'v', 'k']] = df_coords[['u', 'v', 'k']].astype(np.int64)

~/anaconda3/envs/ox/lib/python3.7/site-packages/osmnx/core.py in <listcomp>(.0)
   1199     # first load all the edges' origin and destination coordinates as a
   1200     # dataframe indexed by u, v, key
-> 1201     coords = np.array([[u, v, k, G.nodes[u]['y'], G.nodes[u]['x'], G.nodes[v]['y'], G.nodes[v]['x']] for u, v, k in G.edges(keys=True)])
   1202     df_coords = pd.DataFrame(coords, columns=['u', 'v', 'k', 'u_y', 'u_x', 'v_y', 'v_x'])
   1203     df_coords[['u', 'v', 'k']] = df_coords[['u', 'v', 'k']].astype(np.int64)

KeyError: 'y'
```

**Comment by bradleyswilson** [comment]
Figured this out, it was an issue with having node references on ways that were removed in the osmconvert clipping process. Adding --drop-broken-refs when using osmconvert fixed the issue. 

If you think it would be worth checking for this when using graph_from_file (or perhaps in overpass_json_from_file), I'm happy to take a stab at adding it in.

**Comment by awa5114** [comment]
Thanks for following up on your own question. I was also facing the same problem https://github.com/gboeing/osmnx/pull/95 but then found your post.

On another note, when I drop the broken refs for the example in my post in https://github.com/gboeing/osmnx/pull/95 I obtain a graph with one node and no edges. 

I checked the resulting .osm file for with and without broken refs and I see that removing broken references completely eradicates all ways and relations from the dataset, leaving only nodes. Do you have any idea why this occurs?

Thanks.

**Comment by bradleyswilson** [comment]
Hard to say without knowing what your .osm data looks like, but I'd guess the ways/relations are defined across the boundary edges of your clipping area (you mentioned it was a very small polygon) and are getting removed. 

Have you tried adding --complete-ways as well? (e.g. osmconvert64 europe.osm.pbf -B=test.poly --complete-ways --drop-broken-refs -o=test.osm) 

That'll make sure the full relations and ways are included when clipping, then you can drop the references to nodes outside the clipping area.

**Comment by antoniogomezalvarado** [comment]
just for anyone out there struggling with this - a working example to cut a subset osm of a larger one overcoming the `KeyError: y` exception
```
`osmconvert --drop-broken-refs -b=-115.2270,36.1360,-115.1165,36.1880 nevada.osm.pbf > lv-center.osm`

**Comment by anisotropi4** [comment]
I ran into the same problem and wrote a two line fix to add_edge_lengths that just checks/drops edges with missing nodes. I'm posting to  checking to see whether this was a known issue (which it looks like it is). The question is, is it worth submitting this as a patch as it is really a data quality issue?

**Comment by antoniogomezalvarado** [comment]
> I ran into the same problem and wrote a two line fix to add_edge_lengths that just checks/drops edges with missing nodes. I'm posting to checking to see whether this was a known issue (which it looks like it is). The question is, is it worth submitting this as a patch as it is really a data quality issue?

I wouldn't use this as a patch since it will hide a problem that you might not be aware of in the beginning.
For example, I didn't have a clue once i've "clipped" an area i'll get broken refs to nodes in the boundary.

**Comment by bradleyswilson** [comment]
> > I ran into the same problem and wrote a two line fix to add_edge_lengths that just checks/drops edges with missing nodes. I'm posting to checking to see whether this was a known issue (which it looks like it is). The question is, is it worth submitting this as a patch as it is really a data quality issue?
> 
> I wouldn't use this as a patch since it will hide a problem that you might not be aware of in the beginning.
> For example, I didn't have a clue once i've "clipped" an area i'll get broken refs to nodes in the boundaries.

I'd think I'd favor a more informative error message. It's definitely a clipping issue not on the osmnx side, but would be helpful to point users in the right direction.

**Comment by gboeing** [comment]
> I'd think I'd favor a more informative error message. It's definitely a clipping issue not on the osmnx side, but would be helpful to point users in the right direction.

I'm open to this proposal.

**Comment by anisotropi4** [comment]
There are then two approaches that come to mind, either a`try/catch`on`KeyError`around the coords generator (line 790 in core.py) on that raises a
`TypeError('Edge with missing nodes likely due to a clipping issue')`or somesuch.

Or accumulate and report on the list of missing notes with an error like:
`TypeError('Edge with missing nodes [123456, 123457...] likely due to a clipping issue')`

I'd prefer the latter as it reports the problem nodes at the cost of an iteration through all edges to identify the missing nodes

I would welcome your thoughts

**Comment by gboeing** [comment]
See proposed fix in #437
